source/language/scope.rst

```diff
@@ -0,0 +1,61 @@
+Variable scope
+==============
+
+This section describes the rules surrounding scoping of variables in OpenQASM.
+
+Global scope
+------------
+
+Identifiers declared as subroutines, gates, and defcals have global scope, 
+*i.e.* they are visible at every scope of the program, and may not be shadowed.
+Subroutines, gates, and defcals may only be declared at the top-level of the
+program. Redeclaration or overloading of subroutines and gates is not allowed.
+Defcals may be overloaded as described in :any:`pulse-gates`.
+
+Non-quantum variables defined at the top level of the program have global scope,
+and are visible at all lower scopes. Globally-scoped variables are
+implicitly ``const`` inside the body of subroutine and gate definitions, though
```

Comment:
Other than discouraging a bad practice, what other motivations led to the implicit const inside the body of subroutine and gate definitions?

Reply:
Mostly the motivation was to limit the ability for subroutines as far as possible, to make that only possible through parameters that are passed as mutable.  Bad practice is a lot of it, with the idea was also to make it clearer to both the compiler and the user about what variables a subroutine may mutate.

With the ability to have "output" variables, it's still possible to mutate global-scope variables, there was just a general consensus within the group that it was better to have them explicitly passed as mutable.

Reply:
I'll also reword the text about "implicitly `const`", since that's actually quite misleading - `const` in OQ3 is a _compile-time_ constant, whereas in this case we mean the variable may have been assigned at run-time, but it's read-only within the subroutine.

Reply:
This behaviour was changed in 07a76c8 (subject to review from the working group), since I believe it's more in-line with another PR we have in development about the meaning of the `const` modifier.